File: tools/mcp/mcp_blender/scripts/render.py
#!/usr/bin/env python3
"""Blender rendering script (single frames, animations, batches).

Run by the MCP server as ``blender --background --python render.py --
<args.json> <job_id>``. Progress is published through job status files and the
final result through the ``MCP_RESULT:`` line (see ``mcp_common.py``).
"""

import os
from pathlib import Path
import sys

import bpy

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from mcp_common import (  # noqa: E402  pylint: disable=wrong-import-position
    ScriptError,
    is_eevee,
    open_project,
    resolve_engine,
    run,
    update_status,
)

# User-facing still formats -> Blender image_settings.file_format identifiers.
IMAGE_FORMATS = {"PNG": "PNG", "JPEG": "JPEG", "JPG": "JPEG", "EXR": "OPEN_EXR", "OPEN_EXR": "OPEN_EXR", "TIFF": "TIFF"}
# User-facing video formats -> FFmpeg container identifiers.
VIDEO_CONTAINERS = {"MP4": "MPEG4", "AVI": "AVI", "MOV": "QUICKTIME", "MKV": "MKV", "WEBM": "WEBM"}


def configure_render(scene, settings, default_engine, default_samples):
    """Apply engine, resolution and sample settings shared by every render op."""
    scene.render.engine = resolve_engine(settings.get("engine"), default_engine)
    resolution = settings.get("resolution")
    if resolution:
        if len(resolution) != 2 or min(resolution) <= 0:
            raise ScriptError(f"resolution must be [width, height] with positive values, got {resolution}")
        scene.render.resolution_x = int(resolution[0])
        scene.render.resolution_y = int(resolution[1])
    if "resolution_percentage" in settings:
        scene.render.resolution_percentage = int(settings["resolution_percentage"])
    samples = int(settings.get("samples", default_samples))
    if scene.render.engine == "CYCLES":
        scene.cycles.samples = samples
        scene.cycles.use_denoising = bool(settings.get("denoise", True))
        scene.cycles.device = "GPU" if settings.get("use_gpu") else "CPU"
    elif is_eevee(scene.render.engine):
        scene.eevee.taa_render_samples = samples
    if "film_transparent" in settings:
        scene.render.film_transparent = bool(settings["film_transparent"])


def set_image_format(scene, fmt):
    """Set a still-image output format; returns the Blender identifier."""
    identifier = IMAGE_FORMATS.get(str(fmt).upper())
    if identifier is None:
        raise ScriptError(f"Unsupported image format '{fmt}'. Supported: PNG, JPEG, EXR, TIFF")
    scene.render.image_settings.file_format = identifier
    return identifier


def require_camera(scene):
    if scene.camera is None:
        raise ScriptError("Scene has no active camera; add one with setup_camera first")


def render_image(args, job_id):
    """Render a single frame to ``output_path``."""
    open_project(args.get("project"))
    scene = bpy.context.scene
    settings = args.get("settings") or {}
    configure_render(scene, settings, "CYCLES", 128)
    set_image_format(scene, settings.get("format", "PNG"))
    require_camera(scene)

    frame = int(args.get("frame", scene.frame_current))
    scene.frame_set(frame)
    output_path = args.get("output_path") or f"/app/outputs/renders/{job_id}.png"
    Path(output_path).parent.mkdir(parents=True, exist_ok=True)
    # The server already picked the right extension; stop Blender appending another.
    scene.render.use_file_extension = False
    scene.render.filepath = output_path

    update_status(job_id, "RUNNING", 10, f"Rendering frame {frame} with {scene.render.engine}")
    bpy.ops.render.render(write_still=True)
    if not Path(output_path).is_file():
        raise ScriptError(f"Render finished but no image was written to {output_path}")
    update_status(job_id, "COMPLETED", 100, "Render complete", output_path=output_path)
    return {"success": True, "output_path": output_path, "frame": frame, "engine": scene.render.engine}


def render_animation(args, job_id):
    """Render a frame range to a video file or a PNG sequence."""
    open_project(args.get("project"))
    scene = bpy.context.scene
    settings = args.get("settings") or {}
    configure_render(scene, settings, "EEVEE", 64)
    require_camera(scene)

    start = int(args.get("start_frame", scene.frame_start))
    end = int(args.get("end_frame", scene.frame_end))
    if end < start:
        raise ScriptError(f"end_frame ({end}) must be >= start_frame ({start})")
    scene.frame_start = start
    scene.frame_end = end
    total = end - start + 1

    output_format = str(settings.get("format", "MP4")).upper()
    output_base = (args.get("output_path") or f"/app/outputs/animations/{job_id}").rstrip("/")
    if output_format == "FRAMES":
        os.makedirs(output_base, exist_ok=True)
        scene.render.image_settings.file_format = "PNG"
        scene.render.filepath = os.path.join(output_base, "frame_####")
        scene.render.use_file_extension = True
        output_path = output_base
    else:
        container = VIDEO_CONTAINERS.get(output_format)
        if container is None:
            raise ScriptError(f"Unsupported animation format '{output_format}'. Supported: MP4, AVI, MOV, MKV, WEBM, FRAMES")
        Path(output_base).parent.mkdir(parents=True, exist_ok=True)
        scene.render.image_settings.file_format = "FFMPEG"
        scene.render.ffmpeg.format = container
        scene.render.ffmpeg.codec = "VP9" if container == "WEBM" else "H264"
        scene.render.ffmpeg.constant_rate_factor = "MEDIUM"
        output_path = f"{output_base}.{output_format.lower()}"
        scene.render.use_file_extension = False
        scene.render.filepath = output_path

    rendered = {"count": 0}

    def on_frame_written(scn, *_):
        rendered["count"] += 1
        progress = min(99, int(rendered["count"] * 100 / total))
        update_status(job_id, "RUNNING", progress, f"Rendered frame {scn.frame_current} ({rendered['count']}/{total})")

    bpy.app.handlers.render_write.append(on_frame_written)
    update_status(job_id, "RUNNING", 0, f"Rendering {total} frames with {scene.render.engine}")
    bpy.ops.render.render(animation=True)

    if output_format != "FRAMES" and not Path(output_path).is_file():
        raise ScriptError(f"Animation render finished but {output_path} was not written")
    update_status(job_id, "COMPLETED", 100, "Animation render complete", output_path=output_path)
    return {"success": True, "output_path": output_path, "frames": total, "format": output_format}


def batch_render(args, job_id):  # noqa: C901
    """Render every (camera, view layer, frame) combination as separate images."""
    open_project(args.get("project"))
    scene = bpy.context.scene
    settings = args.get("settings") or {}
    configure_render(scene, settings, "CYCLES", 128)
    fmt = set_image_format(scene, settings.get("format", "PNG"))
    extension = {"PNG": "png", "JPEG": "jpg", "OPEN_EXR": "exr", "TIFF": "tif"}[fmt]
    scene.render.use_file_extension = False

    frames = [int(f) for f in (args.get("frames") or [scene.frame_current])]
    output_dir = args.get("output_dir") or f"/app/outputs/batch/{job_id}"
    os.makedirs(output_dir, exist_ok=True)

    cameras = []
    for cam_name in args.get("cameras") or []:
        cam = bpy.data.objects.get(cam_name)
        if cam is None or cam.type != "CAMERA":
            raise ScriptError(f"Camera '{cam_name}' not found")
        cameras.append(cam)
    if not cameras:
        require_camera(scene)
        cameras = [scene.camera]

    layer_names = list(args.get("layers") or [])
    for name in layer_names:
        if name not in scene.view_layers:
            available = ", ".join(v.name for v in scene.view_layers)
            raise ScriptError(f"View layer '{name}' not found. Available: {available}")
    layers = layer_names or [None]

    total = len(frames) * len(cameras) * len(layers)
    done = 0
    output_files = []
    update_status(job_id, "RUNNING", 0, f"Starting batch render: {total} images")
    original_use = {v.name: v.use for v in scene.view_layers}
    try:
        for layer in layers:
            if layer is not None:
                for view_layer in scene.view_layers:
                    view_layer.use = view_layer.name == layer
            for cam in cameras:
                scene.camera = cam
                for frame in frames:
                    scene.frame_set(frame)
                    parts = [cam.name.replace(" ", "_")]
                    if layer is not None:
                        parts.append(layer.replace(" ", "_"))
                    parts.append(f"frame_{frame:04d}")
                    output_file = os.path.join(output_dir, "_".join(parts) + f".{extension}")
                    scene.render.filepath = output_file
                    bpy.ops.render.render(write_still=True)
                    output_files.append(output_file)
                    done += 1
                    update_status(job_id, "RUNNING", min(99, int(done * 100 / total)), f"Rendered {done}/{total}")
    finally:
        for view_layer in scene.view_layers:
            view_layer.use = original_use.get(view_layer.name, True)

    update_status(job_id, "COMPLETED", 100, f"Batch render complete: {len(output_files)} images", output_path=output_dir)
    return {"success": True, "output_path": output_dir, "output_files": output_files}


def main():
    """Dispatch the requested operation (see mcp_common.run)."""
    run({"render_image": render_image, "render_animation": render_animation, "batch_render": batch_render})


if __name__ == "__main__":
    main()

File: tools/mcp/mcp_blender/src/tools/render.rs
//! Rendering tools. All of them run as async jobs.

use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

use crate::server::{
    BlenderTool, Ctx, ToolError, ToolOutput, check_name, check_range, check_resolution, enum_prop,
    project_prop,
};
use crate::types::{ImageFormat, RenderEngine, VideoFormat};

/// Most frames a single animation/batch job may render.
const MAX_FRAMES: i64 = 100_000;

/// Render settings shared by the render tools (format type varies).
#[derive(Debug, Deserialize, Serialize)]
#[serde(bound(deserialize = "F: Deserialize<'de>"))]
pub struct RenderSettings<F> {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    resolution: Option<[u32; 2]>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    resolution_percentage: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    samples: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    engine: Option<RenderEngine>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    format: Option<F>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    use_gpu: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    denoise: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    film_transparent: Option<bool>,
}

impl<F> Default for RenderSettings<F> {
    fn default() -> Self {
        Self {
            resolution: None,
            resolution_percentage: None,
            samples: None,
            engine: None,
            format: None,
            use_gpu: None,
            denoise: None,
            film_transparent: None,
        }
    }
}

impl<F> RenderSettings<F> {
    fn validate(&self) -> Result<(), ToolError> {
        if let Some(res) = &self.resolution {
            check_resolution(res)?;
        }
        if let Some(pct) = self.resolution_percentage {
            check_range("settings.resolution_percentage", pct, 1, 100)?;
        }
        if let Some(samples) = self.samples {
            check_range("settings.samples", samples, 1, 65536)?;
        }
        Ok(())
    }
}

fn settings_schema(
    formats: &[&str],
    default_format: &str,
    default_engine: &str,
    default_samples: u32,
) -> Value {
    json!({
        "type": "object",
        "description": "Render settings (unset values keep the project's own settings)",
        "properties": {
            "resolution": { "type": "array", "items": { "type": "integer", "minimum": 1, "maximum": 16384 }, "minItems": 2, "maxItems": 2, "description": "[width, height]" },
            "resolution_percentage": { "type": "integer", "minimum": 1, "maximum": 100 },
            "samples": { "type": "integer", "minimum": 1, "default": default_samples },
            "engine": enum_prop(RenderEngine::ALL, Some(default_engine), "CYCLES works everywhere; EEVEE/Workbench need an OpenGL/EGL-capable host"),
            "format": enum_prop(formats, Some(default_format), "Output format"),
            "use_gpu": { "type": "boolean", "default": false, "description": "Cycles: render on the GPU" },
            "denoise": { "type": "boolean", "default": true, "description": "Cycles: denoise the result" },
            "film_transparent": { "type": "boolean", "description": "Transparent background" }
        }
    })
}

fn job_started(job_type: &str, job_id: uuid::Uuid, output_path: &str, extra: Value) -> Value {
    let mut out = json!({
        "success": true,
        "job_id": job_id.to_string(),
        "status": "QUEUED",
        "output_path": output_path,
        "message": format!("{job_type} job queued; poll get_job_status (use wait_seconds) and fetch get_job_result"),
    });
    if let (Some(out), Some(extra)) = (out.as_object_mut(), extra.as_object()) {
        out.extend(extra.clone());
    }
    out
}

/// `render_image` arguments.
#[derive(Debug, Deserialize)]
pub struct RenderImageArgs {
    project: String,
    #[serde(default = "one")]
    frame: i64,
    #[serde(default)]
    settings: RenderSettings<ImageFormat>,
}

fn one() -> i64 {
    1
}

/// Render one frame to an image.
pub struct RenderImage;

#[async_trait]
impl BlenderTool for RenderImage {
    type Args = RenderImageArgs;
    const NAME: &'static str = "render_image";
    const DESCRIPTION: &'static str = "Render a single frame to an image (PNG, JPEG, EXR or TIFF) using the scene's \
active camera. Runs as an async job: returns a job_id immediately; use get_job_status (wait_seconds) and \
get_job_result. The image is written to <output_dir>/renders/<job_id>.<ext>.";

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": {
                "project": project_prop(),
                "frame": { "type": "integer", "default": 1, "description": "Frame to render" },
                "settings": settings_schema(ImageFormat::ALL, "PNG", "CYCLES", 128)
            },
            "required": ["project"]
        })
    }

    async fn run(ctx: &Ctx, args: Self::Args) -> ToolOutput {
        let project = ctx.project(&args.project)?;
        args.settings.validate()?;
        check_range("frame", args.frame, -MAX_FRAMES, MAX_FRAMES * 10)?;
        let format = args.settings.format.unwrap_or(ImageFormat::Png);

        let renders = ctx.config.output_dir.join("renders");
        let output_for = |id: uuid::Uuid| renders.join(format!("{id}.{}", format.extension()));
        let project_path = project.path.to_string_lossy().to_string();
        let frame = args.frame;
        let settings = args.settings;
        let id = ctx.start_job(Self::NAME, "render.py", &project, false, |id| {
            json!({
                "operation": "render_image",
                "project": project_path,
                "frame": frame,
                "settings": settings,
                "output_path": output_for(id).to_string_lossy(),
            })
        });
        let output_path = output_for(id);
        Ok(job_started(
            "Render",
            id,
            &output_path.to_string_lossy(),
            json!({ "frame": args.frame, "format": format }),
        ))
    }
}

/// `render_animation` arguments.
#[derive(Debug, Deserialize)]
pub struct RenderAnimationArgs {
    project: String,
    #[serde(default = "one")]
    start_frame: i64,
    #[serde(default = "default_end")]
    end_frame: i64,
    #[serde(default)]
    settings: RenderSettings<VideoFormat>,
}

fn default_end() -> i64 {
    250
}

/// Render a frame range to a video or image sequence.
pub struct RenderAnimation;

#[async_trait]
impl BlenderTool for RenderAnimation {
    type Args = RenderAnimationArgs;
    const NAME: &'static str = "render_animation";
    const DESCRIPTION: &'static str = "Render a frame range to a video (MP4/H.264, AVI, MOV, MKV, WEBM/VP9) or a PNG \
sequence (FRAMES). Runs as an async job with per-frame progress; use get_job_status (wait_seconds) and get_job_result. \
Output goes to <output_dir>/animations/<id>.<ext> (or that directory for FRAMES).";

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": {
                "project": project_prop(),
                "start_frame": { "type": "integer", "default": 1 },
                "end_frame": { "type": "integer", "default": 250 },
                "settings": settings_schema(VideoFormat::ALL, "MP4", "BLENDER_EEVEE", 64)
            },
            "required": ["project"]
        })
    }

    async fn run(ctx: &Ctx, args: Self::Args) -> ToolOutput {
        let project = ctx.project(&args.project)?;
        args.settings.validate()?;
        if args.end_frame < args.start_frame {
            return Err(ToolError::invalid(format!(
                "end_frame ({}) must be >= start_frame ({})",
                args.end_frame, args.start_frame
            )));
        }
        let frames = args.end_frame - args.start_frame + 1;
        check_range("frame count", frames, 1, MAX_FRAMES)?;
        let format = args.settings.format.unwrap_or(VideoFormat::Mp4);

        let animations = ctx.config.output_dir.join("animations");
        let project_path = project.path.to_string_lossy().to_string();
        let (start, end, settings) = (args.start_frame, args.end_frame, args.settings);
        let id = ctx.start_job(Self::NAME, "render.py", &project, false, |id| {
            json!({
                "operation": "render_animation",
                "project": project_path,
                "start_frame": start,
                "end_frame": end,
                "settings": settings,
                "output_path": animations.join(id.to_string()).to_string_lossy(),
            })
        });
        let base = animations
            .join(id.to_string())
            .to_string_lossy()
            .to_string();
        let output_path = if format == VideoFormat::Frames {
            base
        } else {
            format!("{base}.{}", format.as_str().to_ascii_lowercase())
        };
        Ok(job_started(
            "Animation render",
            id,
            &output_path,
            json!({ "frames": frames, "format": format }),
        ))
    }
}

/// `batch_render` arguments.
#[derive(Debug, Deserialize)]
pub struct BatchRenderArgs {
    project: String,
    #[serde(default)]
    frames: Option<Vec<i64>>,
    #[serde(default)]
    cameras: Vec<String>,
    #[serde(default)]
    layers: Vec<String>,
    #[serde(default)]
    settings: RenderSettings<ImageFormat>,
}

/// Render every combination of frames, cameras and view layers.
pub struct BatchRender;

#[async_trait]
impl BlenderTool for BatchRender {
    type Args = BatchRenderArgs;
    const NAME: &'static str = "batch_render";
    const DESCRIPTION: &'static str = "Render every combination of frames x cameras x view layers as separate images \
(named <camera>[_<layer>]_frame_NNNN.<ext>) into <output_dir>/batch/<id>/. Defaults: frame 1, the active camera, \
all enabled view layers composited. Runs as an async job; get_job_result lists output_files.";

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": {
                "project": project_prop(),
                "frames": { "type": "array", "items": { "type": "integer" }, "default": [1], "description": "Frames to render" },
                "cameras": { "type": "array", "items": { "type": "string" }, "description": "Camera objects to render from (default: active camera)" },
                "layers": { "type": "array", "items": { "type": "string" }, "description": "View layers to render one at a time (default: all together)" },
                "settings": settings_schema(ImageFormat::ALL, "PNG", "CYCLES", 128)
            },
            "required": ["project"]
        })
    }

    async fn run(ctx: &Ctx, args: Self::Args) -> ToolOutput {
        let project = ctx.project(&args.project)?;
        args.settings.validate()?;
        let frames = args.frames.unwrap_or_else(|| vec![1]);
        if frames.is_empty() {
            return Err(ToolError::invalid("frames must not be empty"));
        }
        for camera in &args.cameras {
            check_name("cameras", camera)?;
        }
        for layer in &args.layers {
            check_name("layers", layer)?;
        }
        let total = frames.len() as i64
            * args.cameras.len().max(1) as i64
            * args.layers.len().max(1) as i64;
        check_range("total images (frames x cameras x layers)", total, 1, 10_000)?;

        let batch = ctx.config.output_dir.join("batch");
        let project_path = project.path.to_string_lossy().to_string();
        let (cameras, layers, settings) = (args.cameras, args.layers, args.settings);
        let id = ctx.start_job(Self::NAME, "render.py", &project, false, |id| {
            json!({
                "operation": "batch_render",
                "project": project_path,
                "frames": frames,
                "cameras": cameras,
                "layers": layers,
                "settings": settings,
                "output_dir": batch.join(id.to_string()).to_string_lossy(),
            })
        });
        let output_dir = batch.join(id.to_string());
        Ok(job_started(
            "Batch render",
            id,
            &output_dir.to_string_lossy(),
            json!({ "images": total }),
        ))
    }
}
